request_test.go

```diff
@@ -1601,3 +1603,171 @@ func TestRequestErrorConflictMultipart(t *testing.T) {
 	req3.WithFileBytes("a", "a", []byte("a"))
 	req3.chain.assertFailed(t)
 }
+
+func TestRequestRedirect(t *testing.T) {
+	reporter := newMockReporter(t)
+
+	t.Run("dont follow redirects policy", func(t *testing.T) {
+		trans := newMockTransportRedirect(t, func(r *http.Request) bool {
+			return assert.Nil(t, r.GetBody)
+		})
+
+		client := &http.Client{
+			Transport: trans,
+		}
+
+		config := Config{
+			Client:   client,
+			Reporter: reporter,
+		}
+
+		req := NewRequest(config, http.MethodGet, "url").
+			WithRedirectPolicy(DontFollowRedirects)
+		req.chain.assertOK(t)
+
+		resp := req.Expect().
+			Status(http.StatusPermanentRedirect)
+		resp.chain.assertOK(t)
+
+		// Should only do round trip 1x
+		assert.Equal(t, uint8(1), trans.tripCount)
+
+		httpClient, _ := req.config.Client.(*http.Client)
+
+		// Should have CheckRedirect set by .Expect()
+		assert.NotNil(t, httpClient.CheckRedirect)
+		assert.Equal(t, http.ErrUseLastResponse, httpClient.CheckRedirect(req.httpReq, nil))
+	})
+
+	t.Run("follow all redirect policy with max redirects 0", func(t *testing.T) {
+		trans := newMockTransportRedirect(t, func(r *http.Request) bool {
+			rc, err := r.GetBody()
+			assert.NoError(t, err)
+
+			b, err := ioutil.ReadAll(rc)
+			assert.NoError(t, err)
+
+			return assert.Equal(t, "test body", string(b))
+		})
+
+		client := &http.Client{
+			Transport: trans,
+		}
+
+		config := Config{
+			Client:   client,
+			Reporter: reporter,
+		}
+
+		req := NewRequest(config, http.MethodPost, "url").
+			WithRedirectPolicy(FollowAllRedirects).
+			WithMaxRedirects(0).
+			WithText("test body")
+		req.chain.assertOK(t)
+
+		resp := req.Expect()
+		resp.chain.assertFailed(t)
+
+		// Should only do round trip 1x
+		assert.Equal(t, uint8(1), trans.tripCount)
+
+		httpClient, _ := req.config.Client.(*http.Client)
+
+		// Should have CheckRedirect set by .Expect()
+		assert.NotNil(t, httpClient.CheckRedirect)
+		assert.Equal(t, nil, httpClient.CheckRedirect(req.httpReq, make([]*http.Request, 0)))
+		assert.EqualError(t,
+			httpClient.CheckRedirect(req.httpReq, make([]*http.Request, 1)),
+			fmt.Sprintf("stopped after %d redirects", req.maxRedirects),
+		)
+	})
+
+	t.Run("follow all redirect policy with max redirects 1", func(t *testing.T) {
+		trans := newMockTransportRedirect(t, func(r *http.Request) bool {
+			rc, err := r.GetBody()
```

Comment:
hi @gavv i wanna ask something

i still cant figure out why `r.GetBody` is nil in the second redirect (this test `follow all redirect policy with max redirects 1` do 2 redirects)

do you might have any clue?

i think, in the HTTP 308 [mdn web docs](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/308#:~:text=The%20request%20method%20and%20the%20body%20will%20not%20be%20altered) it is written that the payload should be _forwarded_ to the next redirect

Reply:
here is my log when running the test:

```
=== RUN   TestRequestRedirect
=== RUN   TestRequestRedirect/dont_follow_redirects_policy
=== RUN   TestRequestRedirect/follow_all_redirect_policy_with_max_redirects_0
=== CONT  TestRequestRedirect
    mocks_test.go:84: Fail: 
        failed to send http request
          Post "/redirect": stopped after 0 redirects
        
        assertion:
          Request().Expect()
=== RUN   TestRequestRedirect/follow_all_redirect_policy_with_max_redirects_1
--- FAIL: TestRequestRedirect (0.00s)
    --- PASS: TestRequestRedirect/dont_follow_redirects_policy (0.00s)
    --- PASS: TestRequestRedirect/follow_all_redirect_policy_with_max_redirects_0 (0.00s)
    --- FAIL: TestRequestRedirect/follow_all_redirect_policy_with_max_redirects_1 (0.00s)
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x9c6be5]

goroutine 18 [running]:
testing.tRunner.func1.2({0xa4ebe0, 0xfde980})
	/home/rafi/Documents/personal/programming/go/versions/go/src/testing/testing.go:1389 +0x24e
testing.tRunner.func1()
	/home/rafi/Documents/personal/programming/go/versions/go/src/testing/testing.go:1392 +0x39f
panic({0xa4ebe0, 0xfde980})
	/home/rafi/Documents/personal/programming/go/versions/go/src/runtime/panic.go:838 +0x207
github.com/gavv/httpexpect/v2.TestRequestRedirect.func3.1(0xbe22c0?)
	/home/rafi/Documents/personal/foss/github/httpexpect/request_test.go:1687 +0x25
github.com/gavv/httpexpect/v2.(*mockTransportRedirect).RoundTrip(0xc00028a1e0, 0xbe14c0?)
	/home/rafi/Documents/personal/foss/github/httpexpect/request_test.go:1765 +0x98
net/http.send(0xc000292300, {0xbe14c0, 0xc00028a1e0}, {0xac6ee0?, 0x49a301?, 0x0?})
	/home/rafi/Documents/personal/programming/go/versions/go/src/net/http/client.go:252 +0x5d8
net/http.(*Client).send(0xc000282c60, 0xc000292300, {0x7fd16c7ba460?, 0xc0002844d0?, 0x0?})
	/home/rafi/Documents/personal/programming/go/versions/go/src/net/http/client.go:176 +0x9b
net/http.(*Client).do(0xc000282c60, 0xc000292300)
	/home/rafi/Documents/personal/programming/go/versions/go/src/net/http/client.go:725 +0x8f5
net/http.(*Client).Do(0xc000058810?, 0x48c397?)
	/home/rafi/Documents/personal/programming/go/versions/go/src/net/http/client.go:593 +0x19
github.com/gavv/httpexpect/v2.(*Request).sendRequest.func1()
	/home/rafi/Documents/personal/foss/github/httpexpect/request.go:1797 +0x2d
github.com/gavv/httpexpect/v2.(*Request).retryRequest(0xc0002d2000, 0xc0001379b0)
	/home/rafi/Documents/personal/foss/github/httpexpect/request.go:1892 +0x4f9
github.com/gavv/httpexpect/v2.(*Request).sendRequest(0xc0002d2000)
	/home/rafi/Documents/personal/foss/github/httpexpect/request.go:1796 +0x65
github.com/gavv/httpexpect/v2.(*Request).roundTrip(0xc0002d2000)
	/home/rafi/Documents/personal/foss/github/httpexpect/request.go:1703 +0x105
github.com/gavv/httpexpect/v2.(*Request).Expect(0xc0002d2000)
	/home/rafi/Documents/personal/foss/github/httpexpect/request.go:1664 +0x166
github.com/gavv/httpexpect/v2.TestRequestRedirect.func3(0xc0002d0000)
	/home/rafi/Documents/personal/foss/github/httpexpect/request_test.go:1712 +0x1fa
testing.tRunner(0xc0002d0000, 0xc0002843e0)
	/home/rafi/Documents/personal/programming/go/versions/go/src/testing/testing.go:1439 +0x102
created by testing.(*T).Run
	/home/rafi/Documents/personal/programming/go/versions/go/src/testing/testing.go:1486 +0x35f
FAIL	github.com/gavv/httpexpect/v2	0.010s
FAIL
make: *** [Makefile:37: ut] Error 1
```

many thanks!

Reply:
my Makefile:

```
ut:
	clear && \
		go test -v -run TestRequestRedirect github.com/gavv/httpexpect/v2
```

Reply:
after some debugging

CMIIW it looks like `.GetBody` is not set for server request, such as redirect

as written in the code ([src/net/http/client.go](https://cs.opensource.google/go/go/+/refs/tags/go1.19.3:src/net/http/client.go;l=663;drc=d5de62df152baf4de6e9fe81933319b86fd95ae4))

and as written in the docs ([net/http/Request.GetBody](https://pkg.go.dev/net/http#:~:text=//%20For%20server%20requests%2C%20it%20is%20unused.))

Reply:
i have updated the assertion to use `TeeReader` instead to assert the `r.Body`. wdyt @gavv ? thanks!

https://github.com/gavv/httpexpect/pull/172/commits/7e90c0c3b46688963339bfbb607351efe2d52865

Reply:
Yes, exactly. The convention is the following:

* we create http.Request that has both Body and GetBody
* we pass it to http.Client
* http.Client passes request to RoundTripper as is
* if roundtripper returns redirection response, and CheckRedirect reports that redirect should be followed, http.Client creates a new request
* in new request, http.Client doesn't set GetBody; instead, it calls GetBody on original request and sets Body on new request
* the loop with redirects is repeated until CheckRedirect tells to stop

So yes, in round tripper we should check Body, not GetBody. GetBody is only for http.Client, not for roundtripper.

Relates lines:

* https://cs.opensource.google/go/go/+/refs/tags/go1.19.3:src/net/http/client.go;l=627-628
* https://cs.opensource.google/go/go/+/refs/tags/go1.19.3:src/net/http/client.go;l=652-668

Reply:
I think we should change all tests to use r.Body, and ignore r.GetBody in RoundTripper (it first call it will be non-nil, in other calls it will be nil, but that's not really important and can be considered an implementation detail).

On the other hand, I think we should still do a check that .Expect() sets GetBody in the *original* request (req.httpReq.GetBody). We can also call that GetBody and check that the body is correct.

Reply:
BTW I think there is no real need for TeeReader. I think http.Client expects that RoundTripper will read the body and close the reader. And each round trip, http.Client will set Body to fresh reader returned by GetBody.

Reply:
Ideally we should also add a test for success case with FollowAllRedirects: when a few redirects are followed, limit is NOT reached, Expect() succeeds, and response has status 200. For that, we'll need to teach RoundTripper to return 200 instead of redirect eventually.

Reply:
Also ideally we should cover FollowRedirectsWithoutBody (cases when request has and does not have body).

Reply:
Regarding understanding of GetBody. You can think about it as an optional thing needed solely by http.Client, solely for redirect support.

Normally request doesn't have GetBody (both on client and server). And even if it has one, normally it would be ignored.

However in one special case, when http.Client sent request, and received response with redirect status, and decided to follow redirect, and when request has body (e.g. it's POST or PUT), then http.Client uses GetBody to "rollback" Body to its original value. Without GetBody it wont be possible because Body's contract is that it's read once and then closed, without any possibility to "rollback".

And in this special case http.Client performs this "rollback" magic with Body and then passes to RoundTripper a normal request with Body and without GetBody and RoundTripper works as usual.

Reply:
> Ideally we should also add a test for success case with FollowAllRedirects: when a few redirects are followed, limit is NOT reached, Expect() succeeds, and response has status 200. For that, we'll need to teach RoundTripper to return 200 instead of redirect eventually.

@gavv  does it mean that our round tripper implementation (e.g. `newMockTransportRedirect.RoundTrip`) should be able to do `RoundTrip` with different HTTP response?

something like

initial request --> RoundTrip returns HTTP 308 --> redirect --> RoundTrip returns HTTP 200 --> response OK

Reply:
Yep, ideally a few redirects and then HTTP 200.

A simple way to do it is to add a redirect limit, until counter reaches this limit, we'll return redirect, and after limit is reached we return 200.

Reply:
This is how it normally works in practice. http.Client follows a few redirects and finally gets desired response. So it makes sense to cover it in unit tests.

Reply:
i see, okok

i will limit the scope of this PR into:

- add round tripper implementation that can have various HTTP response in unit tests
- add unit tests for redirects

regarding retry mechanism will not be included in this PR

thank you for the feeback, will do the update

Reply:
Great, thank you.